### PR TITLE
Fixed #28687 -- Added a 'Not Empty' option to admin's related filter.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/28687

### CODE REVIEW

**Comment by hramezani** [comment]
Hi @carltongibson,

Thank you so much for the comments.

- I moved it next to `Empty` option.
- I added relase note for it.
- I added test for it.

In my point of view there is no need to an equivalent of `self.empty_value_display`, but please make final decision about it.

File: tests/admin_filters/tests.py [hunk 1]
```diff
@@ -472,6 +472,16 @@ def test_allvaluesfieldlistfilter(self):
         self.assertIs(choices[-1]['selected'], True)
         self.assertEqual(choices[-1]['query_string'], '?year__isnull=True')
 
+        request = self.request_factory.get('/', {'year__isnull': 'False'})
+        changelist = modeladmin.get_changelist_instance(request)
+
+        # Make sure the correct choice is selected
```
Comment (carltongibson) [inline tests/admin_filters/tests.py]: This could be more specific/clearer.

File: tests/admin_filters/tests.py [hunk 2]
```diff
@@ -472,6 +472,16 @@ def test_allvaluesfieldlistfilter(self):
...
+        filterspec = changelist.get_filters(request)[0][0]
+        self.assertEqual(filterspec.title, 'year')
+        choices = list(filterspec.choices(changelist))
+        self.assertIs(choices[-2]['selected'], False)
```
Comment (carltongibson) [inline tests/admin_filters/tests.py]: Is this right? Doesn't this say "The next-to-last choice is **not** selected"?

**Comment by hramezani** [comment]
@carltongibson thanks again.

I wrote a better comment for the test. and also change the `selected` condition. so the test selected assert changed too.

**Comment by hramezani** [comment]
@carltongibson , thanks.
I added the option for `RelatedFieldListFilter`.

File: docs/releases/2.1.txt [hunk 3]
```diff
@@ -67,6 +67,8 @@ Minor features
 * :meth:`.InlineModelAdmin.has_add_permission` is now passed the parent object
   as the second positional argument, ``obj``.
 
+* ``Not Empty`` option added to admin's related filter.
```
Comment (carltongibson) [inline docs/releases/2.1.txt]: `...related and all values filters.`

**Comment by hramezani** [comment]
@carltongibson, yes I will add this as soon as possible.

**Comment by hramezani** [comment]
@carltongibson , I added `get_not_empty_value_display()`.

**Comment by hramezani** [comment]
@carltongibson is there anything left?

**Comment by carltongibson** [comment]
Hi @hramezani — Only that I need to have another look. (I will do that this week.) Thanks.

**Comment by hramezani** [comment]
Hi @carltongibson, 
Thank you so much for the comments.
I added the tests and also changed release note.

**Comment by hramezani** [comment]
Hi again @carltongibson,

> Whilst I can see the benefit of the feature, I have a slight reservation about adding yet another admin option for this. I wonder if it wouldn't be better fully encapsulated in a filter subclass?

can you help me about this paragraph?

**Comment by carltongibson** [comment]
Hi @hramezani. 

> can you help me about this paragraph?

Sure. 

What I mean is that `list_filter` [accepts `FieldListFilter` subclasses directly]. There are already **many** admin options. A part of me thinks adding an extra one here is _just more noise_ in the API, for (maybe) not so much benefit. That part of me would be inclined to say that users should create their own `FieldListFilter` subclass if they want this extra behaviour and specify it directly. That way we don't add extra clutter for everyone. 

However, the PR itself itself looks OK. I will add my reservation to the ticket and mark it RFC to invite another review. Thanks for your effort.

**Comment by hramezani** [comment]
@timgraham I added missed `versionadded`.

**Comment by timgraham** [comment]
Did you see my comment on the ticket? I also agree with Carlton's assessment that it's not clear this is a feature that everybody wants. It would be interesting to see what's required to implement this in a subclass.

**Comment by hramezani** [comment]
@timgraham, could you help me, please? I can't understand what is your mean. how can I implement it in a subclass? thank you

**Comment by timgraham** [comment]
The idea is that maybe it would be enough for anyone who wants this feature to subclass RelatedFieldListFilter and override the `choices()` method to add the extra choice. That way we don't have to make any changes in Django and people who don't want this option won't be bothered by it.

**Comment by hramezani** [comment]
ok, I got it. it seems you can close this pull request

**Comment by timgraham** [comment]
Okay, I closed the ticket as wontfix.

**Pull request closed without merging**